x-pack/plugins/case/server/services/user_actions/helpers.ts

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { SavedObject, SavedObjectsUpdateResponse } from 'kibana/server';
+import { get } from 'lodash';
+
+import {
+  CaseUserActionAttributes,
+  UserAction,
+  UserActionField,
+  CaseAttributes,
+  User,
+} from '../../../common/api';
+import { isTwoArraysDifference } from '../../routes/api/cases/helpers';
+import { UserActionItem } from '.';
+import { CASE_SAVED_OBJECT, CASE_COMMENT_SAVED_OBJECT } from '../../saved_object_types';
+
+export const transformNewUserAction = ({
+  actionField,
+  action,
+  actionAt,
+  email,
+  full_name,
+  newValue = null,
+  oldValue = null,
+  username,
+}: {
+  actionField: UserActionField;
+  action: UserAction;
+  actionAt: string;
+  email?: string;
+  full_name?: string;
+  newValue?: string | null;
+  oldValue?: string | null;
+  username: string;
+}): CaseUserActionAttributes => ({
+  action_field: actionField,
+  action,
+  action_at: actionAt,
+  action_by: { email, full_name, username },
+  new_value: newValue,
+  old_value: oldValue,
+});
+
+interface BuildCaseUserAction {
+  action: UserAction;
+  actionAt: string;
+  actionBy: User;
+  caseId: string;
+  fields: UserActionField | unknown[];
+  newValue?: string | unknown;
+  oldValue?: string | unknown;
+}
+
+interface BuildCommentUserActionItem extends BuildCaseUserAction {
+  commentId: string;
+}
+
+export const buildCommentUserActionItem = ({
+  action,
+  actionAt,
+  actionBy,
+  caseId,
+  commentId,
+  fields,
+  newValue,
+  oldValue,
+}: BuildCommentUserActionItem): UserActionItem => ({
+  attributes: transformNewUserAction({
+    actionField: fields as UserActionField,
+    action,
+    actionAt,
+    ...actionBy,
+    newValue: newValue as string,
+    oldValue: oldValue as string,
+  }),
+  references: [
+    {
+      type: CASE_SAVED_OBJECT,
+      name: `associated-${CASE_SAVED_OBJECT}`,
+      id: caseId,
+    },
+    {
+      type: CASE_COMMENT_SAVED_OBJECT,
+      name: `associated-${CASE_COMMENT_SAVED_OBJECT}`,
+      id: commentId,
+    },
+  ],
+});
+
+export const buildCaseUserActionItem = ({
+  action,
+  actionAt,
+  actionBy,
+  caseId,
+  fields,
+  newValue,
+  oldValue,
+}: BuildCaseUserAction): UserActionItem => ({
+  attributes: transformNewUserAction({
+    actionField: fields as UserActionField,
+    action,
+    actionAt,
+    ...actionBy,
+    newValue: newValue as string,
+    oldValue: oldValue as string,
+  }),
+  references: [
+    {
+      type: CASE_SAVED_OBJECT,
+      name: `associated-${CASE_SAVED_OBJECT}`,
+      id: caseId,
+    },
+  ],
+});
+
+interface CompareArray {
+  addedItems: string[];
+  deletedItems: string[];
+}
+const compareArray = ({
```

Comment:
I think is better to use a `Map` to reduce the time complexity from `O(n^2)` to `O(n)` sacrificing `O(n)` space

Reply:
Like we discussed together,  it will make a difference if we are dealing with a lot of items in the array but we are not